sentry-tower/src/http.rs

```diff
@@ -0,0 +1,146 @@
+#[derive(Clone, Default)]
+pub struct SentryHttpLayer {
+    start_transaction: bool,
+}
+
+impl SentryHttpLayer {
+    /// Creates a new Layer that only logs Request Headers.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new Layer that also starts a new performance monitoring transaction.
+    pub fn with_transaction() -> Self {
+        Self {
+            start_transaction: true,
+        }
+    }
+}
+
+/// Tower Service that logs Http Request Headers.
+///
+/// The Service can also optionally start a new performance monitoring transaction
+/// based on incoming distributed tracing headers.
+#[derive(Clone)]
+pub struct SentryHttpService<S> {
+    service: S,
+    start_transaction: bool,
+}
+
+impl<S> Layer<S> for SentryHttpLayer {
+    type Service = SentryHttpService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        Self::Service {
+            service,
+            start_transaction: self.start_transaction,
+        }
+    }
+}
+
+/// The Future returned from [`SentryHttpService`]
+#[pin_project::pin_project]
+pub struct SentryHttpFuture<F> {
+    transaction: Option<(
+        sentry_core::TransactionOrSpan,
+        Option<sentry_core::TransactionOrSpan>,
+    )>,
+    #[pin]
+    future: F,
+}
+
+impl<F> Future for SentryHttpFuture<F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let slf = self.project();
+        match slf.future.poll(cx) {
+            Poll::Ready(res) => {
+                if let Some((transaction, parent_span)) = slf.transaction.take() {
+                    transaction.finish();
+                    sentry_core::configure_scope(|scope| scope.set_span(parent_span));
+                }
+                Poll::Ready(res)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl<S, Body> Service<Request<Body>> for SentryHttpService<S>
+where
+    S: Service<Request<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = SentryHttpFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        let transaction = sentry_core::configure_scope(|scope| {
+            let sentry_req = sentry_core::protocol::Request {
+                method: Some(request.method().to_string()),
+                url: request.uri().to_string().parse().ok(),
+                headers: request
+                    .headers()
+                    .into_iter()
+                    .map(|(header, value)| {
+                        (
+                            header.to_string(),
+                            value.to_str().unwrap_or_default().into(),
+                        )
+                    })
+                    .collect(),
```

Comment:
this seems expensive.  do we really need to copy all the headers?  can we be selective, i.e. only copy a few headers we need?  can we avoid copying?

Reply:
The `protocol::Request` owns all of its data (and so does `protocol::Event` [the `Cow` in it are a lie basically, or rather only `'static` ones are allowed -_-])
So we can’t avoid the copies, and I’m not sure we should be opinionated what kind of headers we log. Also, the `actix` integration does the same already.